### ISSUE TITLE
ConcurrencyToken (RowVersion/Timestamp columns) are not checked during bulk operations

### ISSUE THREAD

**Issue by isak1080** [zzzprojects/EntityFramework-Extensions#510]
### Description
_This is related to #509 but I wanted to open a separate issue to clearly document what works and what not (that other issue is more about documentation/example being wrong)_

It seems that most of the bulk operations in EF Core completely ignore [concurrency tokens](https://learn.microsoft.com/en-us/ef/core/modeling/concurrency). In EF 6, this used to work (as described by the [zzz project docs](https://entityframework-extensions.net/concurrency)).

Specifically, I'd expect the following calls to refuse to update entities that have a RowVersion value that is older/different than the one in the database:

* `BulkUpdate()` 
* `BulkDelete()` 
* `BulkSaveChanges()`
including their `...Async()` counterparts of course.

Of the above, only `BulkSaveChanges()` actually does this, but it only does it for `UPDATE`, not `DELETE`.

### Code examples

The following works:
```csharp
SomeEntity myEntity = db.Set<SomeEntity>().FirstOrDefault(x => x.Id = SomeId);
// Simulate concurrency by opening another db context and modifying this entity
UpdateMyEntityOnAnotherContext(myEntity.Id)
// Our tracked entity is now old, update it and try to save the changes
myEntity.Foo = "Bar";
// This will do the right thing and throw DbBulkOperationConcurrencyException
db.BulkSaveChanges();
```

The following scenarios does NOT work (they will overwrite/delete stale data):

**Delete stale data using BulkSaveChanges**
```csharp
SomeEntity myEntity = db.Set<SomeEntity>().FirstOrDefault(x => x.Id = SomeId);
// Simulate concurrency by opening another db context and modifying this entity
UpdateMyEntityOnAnotherContext(myEntity.Id)
// Our tracked entity is now old, but BulkSaveChanges will happily perform the delete
db.Remove(myEntity)
db.BulkSaveChanges();
```

**BulkDelete - deletes newer data in db**
```csharp
// Same code as above - myEntity is tracked but not up-to-date compared to db
// Will happily perform the delete
db.BulkDelete(new[]{ myEntity});
```

**BulkUpdate - overwrites with stale data**
```csharp
// Same code as above - myEntity is tracked but not up-to-date compared to db
// Will happily update the delete, silently overwriting changes done on the other context
db.BulkUpdate(new[]{ myEntity});
```
### Further technical details
- EF version: EF Core 6.5.10
- EF Extensions version: Z.EntityFramework.Extensions.EFCore 6.15.2
- Database Provider: SQL Server

**Comment by JonathanMagnan** [comment]
Hello @isak1080 ,

(A copy & paste from the text provided in the #509 issue)

With the latest version, we updated how we handle concurrency. See the following documentation: https://entityframework-extensions.net/concurrency

You will find an example of how to manage concurrency with `BulkSaveChanges`.

Due to backward compatibility, you will need to enable concurrency for operations such as `BulkDelete`, `BulkUpdate`, and `BulkMerge` with the following options: `EnableConcurrencyForBulkOperation = true` . We cannot enable it by default as we might break many codes currently in production.

Let me know if you have any questions about it.

Best Regards,

Jon

**Comment by JonathanMagnan** [comment]
Hello @isak1080 ,

We added the recent concurrency changes in the the v6.16.1

**Comment by isak1080** [comment]
Hi, sorry for the delay in my response.
I've now run my tests with the latest version (used 7.0 preview) and it works fine *except in one case*: when doing a delete and then using `BulkSaveChanges`

i.e. this scenario will (incorrectly) delete the entity when it should instead throw an exception:

```cs
SomeEntity myEntity = db.Set<SomeEntity>().FirstOrDefault(x => x.Id = SomeId);
// Simulate concurrency by opening another db context and modifying this entity
UpdateMyEntityOnAnotherContext(myEntity.Id)
// Our tracked entity is now old, but BulkSaveChanges will happily perform the delete
db.Remove(myEntity)
// Doesn't matter if specifying opt.AllowConcurrency = true, or opt.EnableConcurrencyForBulkOperation = true
// The entity will be deleted regardless
db.BulkSaveChanges(opt => opt.AllowConcurrency = true);
```

**Comment by JonathanMagnan** [comment]
Thank you @isak1080 for reporting, we will look at this.

**Comment by JonathanMagnan** [comment]
Hello @isak1080 ,

A new version is now available (v6.18.0 for EF Core 6): https://github.com/zzzprojects/EntityFramework-Extensions/releases/tag/7.18.0

We added a breaking change in our library to correctly throw an error when a deleted entity should throw a concurrency error.

So it should now work as you should expect it to work for `BulkSaveChanges`.

Let me know if you can test this version if you get the expected behavior now.

Best Regards,

Jon

**Comment by isak1080** [comment]
Great. Works fine - all my tests now pass.

Thanks for fixing this
